### PR TITLE
Graph.hpp: fixes segmenation fault for degree functions

### PR DESCRIPTION
**Pull request by PetholzA**
fixes {degree, degreeOut, degreeIn} functions when called with an empty Graph

### CODE REVIEW

**Comment by fabratu** [comment]
Thanks for the PR. Please mention the linked issue (using `#` + the number of the issue).

**Comment by PetholzA** [comment]
#1007

File: include/networkit/graph/Graph.hpp [hunk 1]
```diff
@@ -1752,15 +1756,21 @@ class Graph final {
      * @return The number of incoming neighbors.
      * @note If the graph is not directed, the outgoing degree is returned.
      */
-    count degreeIn(node v) const { return directed ? inEdges[v].size() : outEdges[v].size(); }
+    count degreeIn(node v) const { 
+        if (!hasNode(v))
+            throw std::runtime_error("Error, the node does not exist!");
```
Comment (angriman) [inline include/networkit/graph/Graph.hpp:1761]:
```suggestion
    count degreeIn(node v) const { 
        assert(hasNode(v));
```

**Comment by angriman** [comment]
I still disagree with this change. There are examples of algorithms that often use this function, for example [1], [2], [3], and many more. This change would slow them down. I would rather update the documentation of `degree` and `degreeIn` and mention that bound checks are not performed, as done for example by std::vector [4].
As already mentioned, one can always use `assert` to have an informative message if the issue occurs without pessimizing the performance.

[1] https://github.com/networkit/networkit/blob/master/networkit/cpp/centrality/TopCloseness.cpp#L80
[2] https://github.com/networkit/networkit/blob/master/networkit/cpp/centrality/GroupDegree.cpp#LL43C44-L43C44
[3] https://github.com/networkit/networkit/blob/master/networkit/cpp/centrality/KadabraBetweenness.cpp#L645
[4] https://en.cppreference.com/w/cpp/container/vector/operator_at

**Comment by hmeyerhenke** [comment]
I have not looked at the code, but I agree with Eugenio's reasoning. We should avoid performance degradation if other options are available.

**Comment by fabratu** [comment]
In principle, I agree. As already pointed out: For once, we have decided not to integrate checks, if they hamper the performance. Also, `degree`/`degreeIn` are used a lot in algorithms.

Nevertheless, I asked @PetholzA to create a PR to have a discussion about how to find a better / more consistent sweet spot between performance and safety checks. The reasoning is the following: Especially for `Graph.hpp` we include existence checks for nodes in for several functions (also performance critical ones like `forNodes`). Some functions don't have checks at all. We also occasionally use optional parameters to set call functions safely. So there is inconsistency, which create issues for (new) users.

What do you think would be a good option here?
(i) We could of course leave everything as it is. This would lead to problems, especially for Python users, where it is not common to have zero safety checks. Also, this might lead to confusion due to inconsistency.
(ii) Use default parameters. If I have not missed something in the language standard, the C++ function calls can get more messy, since we often have more than one optional parameter. We could add wrapper functions in this case.
(iii) Use asserts. From a user perspective (especially end-users with release builds), I don't think this helps a lot. For development / testing, that makes sense.

**Comment by angriman** [comment]
> In principle, I agree. As already pointed out: For once, we have decided not to integrate checks, if they hamper the performance. Also, `degree`/`degreeIn` are used a lot in algorithms.
> 
> Nevertheless, I asked @PetholzA to create a PR to have a discussion about how to find a better / more consistent sweet spot between performance and safety checks. The reasoning is the following: Especially for `Graph.hpp` we include existence checks for nodes in for several functions (also performance critical ones like `forNodes`). Some functions don't have checks at all. We also occasionally use optional parameters to set call functions safely. So there is inconsistency, which create issues for (new) users.
> 
> What do you think would be a good option here?
> 
> (i) We could of course leave everything as it is. This would lead to problems, especially for Python users, where it is not common to have zero safety checks. Also, this might lead to confusion due to inconsistency. (ii) Use default parameters. If I have not missed something in the language standard, the C++ function calls can get more messy, since we often have more than one optional parameter. We could add wrapper functions in this case. (iii) Use asserts. From a user perspective (especially end-users with release builds), I don't think this helps a lot. For development / testing, that makes sense.

Because NetworKit is a "tool suite for high-performance network analysis", we should not sacrifice performance just to inform users that they are making a mistake in **their** code (e.g., accessing the degree of a deleted or non-existing node). My suggestion is to go with (iii), plus clarify in the documentation that `Graph` does not implement bound checks for performance reasons.

**Comment by angriman** [comment]
@PetholzA thanks for addressing the comments, this looks good to me. It can be merged once the code style issue is fixed.

**Comment by fabratu** [comment]
After having more discussions about this topic - I agree. Let's add asserts where possible for sanity checks and also use the documentation to inform users about possible unsafe behavior of a function. @PetholzA Please add a hint (in both C++ and Python), that the function does not check for existence of a node (and reference `hasNode` to check for it).

Besides that (and the aforementioned style issues), the PR looks good to me.

**Review by angriman — APPROVED** [review]

**Comment by PetholzA** [comment]
note that we also changed the GraphNotebook.ipynb, which included a error

**Review by fabratu — APPROVED** [review]

**Pull request merged**